{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Скачивание файлов близкородственной бактерии"
      ],
      "metadata": {
        "id": "9l4mfp8udffD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "bGAKTNXRdDCt",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ab38482b-2618-475c-fc81-c7c59a28f2c8"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Entrez Direct has been successfully downloaded and installed.\n",
            "\n",
            "In order to complete the configuration process, please execute the following:\n",
            "\n",
            "  echo \"export PATH=\\${PATH}:/root/edirect\" >> ${HOME}/.bashrc\n",
            "\n",
            "or manually edit the PATH variable assignment in your .bashrc file.\n",
            "\n",
            "Would you like to do that automatically now? [y/N]\n",
            "y\n",
            "OK, done.\n",
            "\n",
            "To activate EDirect for this terminal session, please execute the following:\n",
            "\n",
            "export PATH=${PATH}:${HOME}/edirect\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!sh -c \"$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)\""
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gb  >  T_oleivorans_MIL_1.gbk"
      ],
      "metadata": {
        "id": "tdiKn0-mduZ-"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Установка програм "
      ],
      "metadata": {
        "id": "GxMG_d0Pd3lc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%capture\n",
        "!pip install biopython\n",
        "!apt-get update\n",
        "!apt-get install ncbi-blast+\n",
        "!wget https://raw.githubusercontent.com/SpryGorgon/hse22_hw2/main/data/scaffolds.fasta"
      ],
      "metadata": {
        "id": "MhDe8g4Md1xI"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!blastn -help"
      ],
      "metadata": {
        "id": "EbJLs3ZteTv2",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "decc1313-0759-41ce-d814-dc0877e739e4"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "USAGE\n",
            "  blastn [-h] [-help] [-import_search_strategy filename]\n",
            "    [-export_search_strategy filename] [-task task_name] [-db database_name]\n",
            "    [-dbsize num_letters] [-gilist filename] [-seqidlist filename]\n",
            "    [-negative_gilist filename] [-entrez_query entrez_query]\n",
            "    [-db_soft_mask filtering_algorithm] [-db_hard_mask filtering_algorithm]\n",
            "    [-subject subject_input_file] [-subject_loc range] [-query input_file]\n",
            "    [-out output_file] [-evalue evalue] [-word_size int_value]\n",
            "    [-gapopen open_penalty] [-gapextend extend_penalty]\n",
            "    [-perc_identity float_value] [-qcov_hsp_perc float_value]\n",
            "    [-max_hsps int_value] [-xdrop_ungap float_value] [-xdrop_gap float_value]\n",
            "    [-xdrop_gap_final float_value] [-searchsp int_value]\n",
            "    [-sum_stats bool_value] [-penalty penalty] [-reward reward] [-no_greedy]\n",
            "    [-min_raw_gapped_score int_value] [-template_type type]\n",
            "    [-template_length int_value] [-dust DUST_options]\n",
            "    [-filtering_db filtering_database]\n",
            "    [-window_masker_taxid window_masker_taxid]\n",
            "    [-window_masker_db window_masker_db] [-soft_masking soft_masking]\n",
            "    [-ungapped] [-culling_limit int_value] [-best_hit_overhang float_value]\n",
            "    [-best_hit_score_edge float_value] [-window_size int_value]\n",
            "    [-off_diagonal_range int_value] [-use_index boolean] [-index_name string]\n",
            "    [-lcase_masking] [-query_loc range] [-strand strand] [-parse_deflines]\n",
            "    [-outfmt format] [-show_gis] [-num_descriptions int_value]\n",
            "    [-num_alignments int_value] [-line_length line_length] [-html]\n",
            "    [-max_target_seqs num_sequences] [-num_threads int_value] [-remote]\n",
            "    [-version]\n",
            "\n",
            "DESCRIPTION\n",
            "   Nucleotide-Nucleotide BLAST 2.6.0+\n",
            "\n",
            "OPTIONAL ARGUMENTS\n",
            " -h\n",
            "   Print USAGE and DESCRIPTION;  ignore all other parameters\n",
            " -help\n",
            "   Print USAGE, DESCRIPTION and ARGUMENTS; ignore all other parameters\n",
            " -version\n",
            "   Print version number;  ignore other arguments\n",
            "\n",
            " *** Input query options\n",
            " -query <File_In>\n",
            "   Input file name\n",
            "   Default = `-'\n",
            " -query_loc <String>\n",
            "   Location on the query sequence in 1-based offsets (Format: start-stop)\n",
            " -strand <String, `both', `minus', `plus'>\n",
            "   Query strand(s) to search against database/subject\n",
            "   Default = `both'\n",
            "\n",
            " *** General search options\n",
            " -task <String, Permissible values: 'blastn' 'blastn-short' 'dc-megablast'\n",
            "                'megablast' 'rmblastn' >\n",
            "   Task to execute\n",
            "   Default = `megablast'\n",
            " -db <String>\n",
            "   BLAST database name\n",
            "    * Incompatible with:  subject, subject_loc\n",
            " -out <File_Out>\n",
            "   Output file name\n",
            "   Default = `-'\n",
            " -evalue <Real>\n",
            "   Expectation value (E) threshold for saving hits \n",
            "   Default = `10'\n",
            " -word_size <Integer, >=4>\n",
            "   Word size for wordfinder algorithm (length of best perfect match)\n",
            " -gapopen <Integer>\n",
            "   Cost to open a gap\n",
            " -gapextend <Integer>\n",
            "   Cost to extend a gap\n",
            " -penalty <Integer, <=0>\n",
            "   Penalty for a nucleotide mismatch\n",
            " -reward <Integer, >=0>\n",
            "   Reward for a nucleotide match\n",
            " -use_index <Boolean>\n",
            "   Use MegaBLAST database index\n",
            "   Default = `false'\n",
            " -index_name <String>\n",
            "   MegaBLAST database index name (deprecated; use only for old style indices)\n",
            "\n",
            " *** BLAST-2-Sequences options\n",
            " -subject <File_In>\n",
            "   Subject sequence(s) to search\n",
            "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\n",
            "   db_soft_mask, db_hard_mask\n",
            " -subject_loc <String>\n",
            "   Location on the subject sequence in 1-based offsets (Format: start-stop)\n",
            "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\n",
            "   db_soft_mask, db_hard_mask, remote\n",
            "\n",
            " *** Formatting options\n",
            " -outfmt <String>\n",
            "   alignment view options:\n",
            "     0 = Pairwise,\n",
            "     1 = Query-anchored showing identities,\n",
            "     2 = Query-anchored no identities,\n",
            "     3 = Flat query-anchored showing identities,\n",
            "     4 = Flat query-anchored no identities,\n",
            "     5 = BLAST XML,\n",
            "     6 = Tabular,\n",
            "     7 = Tabular with comment lines,\n",
            "     8 = Seqalign (Text ASN.1),\n",
            "     9 = Seqalign (Binary ASN.1),\n",
            "    10 = Comma-separated values,\n",
            "    11 = BLAST archive (ASN.1),\n",
            "    12 = Seqalign (JSON),\n",
            "    13 = Multiple-file BLAST JSON,\n",
            "    14 = Multiple-file BLAST XML2,\n",
            "    15 = Single-file BLAST JSON,\n",
            "    16 = Single-file BLAST XML2,\n",
            "    17 = Sequence Alignment/Map (SAM),\n",
            "    18 = Organism Report\n",
            "   \n",
            "   Options 6, 7, 10 and 17 can be additionally configured to produce\n",
            "   a custom format specified by space delimited format specifiers.\n",
            "   The supported format specifiers for options 6, 7 and 10 are:\n",
            "   \t    qseqid means Query Seq-id\n",
            "   \t       qgi means Query GI\n",
            "   \t      qacc means Query accesion\n",
            "   \t   qaccver means Query accesion.version\n",
            "   \t      qlen means Query sequence length\n",
            "   \t    sseqid means Subject Seq-id\n",
            "   \t sallseqid means All subject Seq-id(s), separated by a ';'\n",
            "   \t       sgi means Subject GI\n",
            "   \t    sallgi means All subject GIs\n",
            "   \t      sacc means Subject accession\n",
            "   \t   saccver means Subject accession.version\n",
            "   \t   sallacc means All subject accessions\n",
            "   \t      slen means Subject sequence length\n",
            "   \t    qstart means Start of alignment in query\n",
            "   \t      qend means End of alignment in query\n",
            "   \t    sstart means Start of alignment in subject\n",
            "   \t      send means End of alignment in subject\n",
            "   \t      qseq means Aligned part of query sequence\n",
            "   \t      sseq means Aligned part of subject sequence\n",
            "   \t    evalue means Expect value\n",
            "   \t  bitscore means Bit score\n",
            "   \t     score means Raw score\n",
            "   \t    length means Alignment length\n",
            "   \t    pident means Percentage of identical matches\n",
            "   \t    nident means Number of identical matches\n",
            "   \t  mismatch means Number of mismatches\n",
            "   \t  positive means Number of positive-scoring matches\n",
            "   \t   gapopen means Number of gap openings\n",
            "   \t      gaps means Total number of gaps\n",
            "   \t      ppos means Percentage of positive-scoring matches\n",
            "   \t    frames means Query and subject frames separated by a '/'\n",
            "   \t    qframe means Query frame\n",
            "   \t    sframe means Subject frame\n",
            "   \t      btop means Blast traceback operations (BTOP)\n",
            "   \t    staxid means Subject Taxonomy ID\n",
            "   \t  ssciname means Subject Scientific Name\n",
            "   \t  scomname means Subject Common Name\n",
            "   \tsblastname means Subject Blast Name\n",
            "   \t sskingdom means Subject Super Kingdom\n",
            "   \t   staxids means unique Subject Taxonomy ID(s), separated by a ';'\n",
            "   \t\t\t (in numerical order)\n",
            "   \t sscinames means unique Subject Scientific Name(s), separated by a ';'\n",
            "   \t scomnames means unique Subject Common Name(s), separated by a ';'\n",
            "   \tsblastnames means unique Subject Blast Name(s), separated by a ';'\n",
            "   \t\t\t (in alphabetical order)\n",
            "   \tsskingdoms means unique Subject Super Kingdom(s), separated by a ';'\n",
            "   \t\t\t (in alphabetical order) \n",
            "   \t    stitle means Subject Title\n",
            "   \tsalltitles means All Subject Title(s), separated by a '<>'\n",
            "   \t   sstrand means Subject Strand\n",
            "   \t     qcovs means Query Coverage Per Subject\n",
            "   \t   qcovhsp means Query Coverage Per HSP\n",
            "   \t    qcovus means Query Coverage Per Unique Subject (blastn only)\n",
            "   When not provided, the default value is:\n",
            "   'qaccver saccver pident length mismatch gapopen qstart qend sstart send\n",
            "   evalue bitscore', which is equivalent to the keyword 'std'\n",
            "   The supported format specifier for option 17 is:\n",
            "   \t        SQ means Include Sequence Data\n",
            "   \t        SR means Subject as Reference Seq\n",
            "   Default = `0'\n",
            " -show_gis\n",
            "   Show NCBI GIs in deflines?\n",
            " -num_descriptions <Integer, >=0>\n",
            "   Number of database sequences to show one-line descriptions for\n",
            "   Not applicable for outfmt > 4\n",
            "   Default = `500'\n",
            "    * Incompatible with:  max_target_seqs\n",
            " -num_alignments <Integer, >=0>\n",
            "   Number of database sequences to show alignments for\n",
            "   Default = `250'\n",
            "    * Incompatible with:  max_target_seqs\n",
            " -line_length <Integer, >=1>\n",
            "   Line length for formatting alignments\n",
            "   Not applicable for outfmt > 4\n",
            "   Default = `60'\n",
            " -html\n",
            "   Produce HTML output?\n",
            "\n",
            " *** Query filtering options\n",
            " -dust <String>\n",
            "   Filter query sequence with DUST (Format: 'yes', 'level window linker', or\n",
            "   'no' to disable)\n",
            "   Default = `20 64 1'\n",
            " -filtering_db <String>\n",
            "   BLAST database containing filtering elements (i.e.: repeats)\n",
            " -window_masker_taxid <Integer>\n",
            "   Enable WindowMasker filtering using a Taxonomic ID\n",
            " -window_masker_db <String>\n",
            "   Enable WindowMasker filtering using this repeats database.\n",
            " -soft_masking <Boolean>\n",
            "   Apply filtering locations as soft masks\n",
            "   Default = `true'\n",
            " -lcase_masking\n",
            "   Use lower case filtering in query and subject sequence(s)?\n",
            "\n",
            " *** Restrict search or results\n",
            " -gilist <String>\n",
            "   Restrict search of database to list of GI's\n",
            "    * Incompatible with:  negative_gilist, seqidlist, remote, subject,\n",
            "   subject_loc\n",
            " -seqidlist <String>\n",
            "   Restrict search of database to list of SeqId's\n",
            "    * Incompatible with:  gilist, negative_gilist, remote, subject,\n",
            "   subject_loc\n",
            " -negative_gilist <String>\n",
            "   Restrict search of database to everything except the listed GIs\n",
            "    * Incompatible with:  gilist, seqidlist, remote, subject, subject_loc\n",
            " -entrez_query <String>\n",
            "   Restrict search with the given Entrez query\n",
            "    * Requires:  remote\n",
            " -db_soft_mask <String>\n",
            "   Filtering algorithm ID to apply to the BLAST database as soft masking\n",
            "    * Incompatible with:  db_hard_mask, subject, subject_loc\n",
            " -db_hard_mask <String>\n",
            "   Filtering algorithm ID to apply to the BLAST database as hard masking\n",
            "    * Incompatible with:  db_soft_mask, subject, subject_loc\n",
            " -perc_identity <Real, 0..100>\n",
            "   Percent identity\n",
            " -qcov_hsp_perc <Real, 0..100>\n",
            "   Percent query coverage per hsp\n",
            " -max_hsps <Integer, >=1>\n",
            "   Set maximum number of HSPs per subject sequence to save for each query\n",
            " -culling_limit <Integer, >=0>\n",
            "   If the query range of a hit is enveloped by that of at least this many\n",
            "   higher-scoring hits, delete the hit\n",
            "    * Incompatible with:  best_hit_overhang, best_hit_score_edge\n",
            " -best_hit_overhang <Real, (>0 and <0.5)>\n",
            "   Best Hit algorithm overhang value (recommended value: 0.1)\n",
            "    * Incompatible with:  culling_limit\n",
            " -best_hit_score_edge <Real, (>0 and <0.5)>\n",
            "   Best Hit algorithm score edge value (recommended value: 0.1)\n",
            "    * Incompatible with:  culling_limit\n",
            " -max_target_seqs <Integer, >=1>\n",
            "   Maximum number of aligned sequences to keep \n",
            "   Not applicable for outfmt <= 4\n",
            "   Default = `500'\n",
            "    * Incompatible with:  num_descriptions, num_alignments\n",
            "\n",
            " *** Discontiguous MegaBLAST options\n",
            " -template_type <String, `coding', `coding_and_optimal', `optimal'>\n",
            "   Discontiguous MegaBLAST template type\n",
            "    * Requires:  template_length\n",
            " -template_length <Integer, Permissible values: '16' '18' '21' >\n",
            "   Discontiguous MegaBLAST template length\n",
            "    * Requires:  template_type\n",
            "\n",
            " *** Statistical options\n",
            " -dbsize <Int8>\n",
            "   Effective length of the database \n",
            " -searchsp <Int8, >=0>\n",
            "   Effective length of the search space\n",
            " -sum_stats <Boolean>\n",
            "   Use sum statistics\n",
            "\n",
            " *** Search strategy options\n",
            " -import_search_strategy <File_In>\n",
            "   Search strategy to use\n",
            "    * Incompatible with:  export_search_strategy\n",
            " -export_search_strategy <File_Out>\n",
            "   File name to record the search strategy used\n",
            "    * Incompatible with:  import_search_strategy\n",
            "\n",
            " *** Extension options\n",
            " -xdrop_ungap <Real>\n",
            "   X-dropoff value (in bits) for ungapped extensions\n",
            " -xdrop_gap <Real>\n",
            "   X-dropoff value (in bits) for preliminary gapped extensions\n",
            " -xdrop_gap_final <Real>\n",
            "   X-dropoff value (in bits) for final gapped alignment\n",
            " -no_greedy\n",
            "   Use non-greedy dynamic programming extension\n",
            " -min_raw_gapped_score <Integer>\n",
            "   Minimum raw gapped score to keep an alignment in the preliminary gapped and\n",
            "   traceback stages\n",
            " -ungapped\n",
            "   Perform ungapped alignment only?\n",
            " -window_size <Integer, >=0>\n",
            "   Multiple hits window size, use 0 to specify 1-hit algorithm\n",
            " -off_diagonal_range <Integer, >=0>\n",
            "   Number of off-diagonals to search for the 2nd hit, use 0 to turn off\n",
            "   Default = `0'\n",
            "\n",
            " *** Miscellaneous options\n",
            " -parse_deflines\n",
            "   Should the query and subject defline(s) be parsed?\n",
            " -num_threads <Integer, >=1>\n",
            "   Number of threads (CPUs) to use in the BLAST search\n",
            "   Default = `1'\n",
            "    * Incompatible with:  remote\n",
            " -remote\n",
            "   Execute search remotely?\n",
            "    * Incompatible with:  gilist, seqidlist, negative_gilist, subject_loc,\n",
            "   num_threads\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Прочитать подробнее можно тут https://www.ncbi.nlm.nih.gov/books/NBK279690/ "
      ],
      "metadata": {
        "id": "fJL5gxFlhnw7"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Предсказание рибосомальной рнк"
      ],
      "metadata": {
        "id": "w5_RP4Yafn_i"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from pprint import pprint\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "mil_1_genome = SeqIO.read(\"/content/T_oleivorans_MIL_1.gbk\", \"genbank\")"
      ],
      "metadata": {
        "id": "zkZ0BC5Ja1n0"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "records=[]\n",
        "for feature in mil_1_genome.features:\n",
        "    if(feature.type=='rRNA' and feature.qualifiers['product'][0][:3] in ['5S ', '16S', '23S']):\n",
        "        record = records.append(SeqRecord(feature.extract(mil_1_genome.seq)))\n",
        "        records[-1].id = feature.qualifiers['locus_tag'][0]\n",
        "        records[-1].description = feature.qualifiers['product'][0]\n",
        "# record.format(\"fasta\")\n",
        "SeqIO.write(records, \"query.fasta\", \"fasta\")\n",
        "# record"
      ],
      "metadata": {
        "id": "gB4iU6j2bBwr",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b6fe2026-42c2-4a53-ee55-3f0690c4b958"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "12"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Ваш код\n",
        "!blastn -query query.fasta -subject scaffolds.fasta -outfmt 7"
      ],
      "metadata": {
        "id": "fng793_WepS2",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2c3aec4e-5f79-4d39-9f24-9fb68c339ade"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_0319 16S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_0319\tscaffold1_cov232\t100.000\t1539\t0\t0\t1\t1539\t2526527\t2528065\t0.0\t2843\n",
            "TOL_0319\tscaffold1_cov232\t99.935\t1539\t1\t0\t1\t1539\t846845\t845307\t0.0\t2837\n",
            "TOL_0319\tscaffold32_cov590\t100.000\t1399\t0\t0\t134\t1532\t1399\t1\t0.0\t2584\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_0322 23S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 5 hits found\n",
            "TOL_0322\tscaffold1_cov232\t100.000\t2887\t0\t0\t1\t2887\t2528316\t2531202\t0.0\t5332\n",
            "TOL_0322\tscaffold1_cov232\t99.965\t2887\t1\t0\t1\t2887\t845056\t842170\t0.0\t5326\n",
            "TOL_0322\tscaffold35_cov581\t100.000\t1535\t0\t0\t1353\t2887\t1815\t281\t0.0\t2835\n",
            "TOL_0322\tscaffold41_cov598\t100.000\t1051\t0\t0\t215\t1265\t1051\t1\t0.0\t1941\n",
            "TOL_0322\tscaffold48_cov623\t100.000\t127\t0\t0\t1\t127\t127\t1\t1.30e-61\t235\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_0323 5S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_0323\tscaffold1_cov232\t100.000\t116\t0\t0\t1\t116\t841988\t841873\t5.67e-57\t215\n",
            "TOL_0323\tscaffold1_cov232\t100.000\t116\t0\t0\t1\t116\t2531384\t2531499\t5.67e-57\t215\n",
            "TOL_0323\tscaffold35_cov581\t100.000\t99\t0\t0\t1\t99\t99\t1\t1.60e-47\t183\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_2453 5S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_2453\tscaffold1_cov232\t98.276\t116\t2\t0\t1\t116\t841988\t841873\t1.23e-53\t204\n",
            "TOL_2453\tscaffold1_cov232\t98.276\t116\t2\t0\t1\t116\t2531384\t2531499\t1.23e-53\t204\n",
            "TOL_2453\tscaffold35_cov581\t98.990\t99\t1\t0\t1\t99\t99\t1\t7.44e-46\t178\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_2454 23S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 5 hits found\n",
            "TOL_2454\tscaffold1_cov232\t100.000\t2887\t0\t0\t1\t2887\t2528316\t2531202\t0.0\t5332\n",
            "TOL_2454\tscaffold1_cov232\t99.965\t2887\t1\t0\t1\t2887\t845056\t842170\t0.0\t5326\n",
            "TOL_2454\tscaffold35_cov581\t100.000\t1535\t0\t0\t1353\t2887\t1815\t281\t0.0\t2835\n",
            "TOL_2454\tscaffold41_cov598\t100.000\t1051\t0\t0\t215\t1265\t1051\t1\t0.0\t1941\n",
            "TOL_2454\tscaffold48_cov623\t100.000\t127\t0\t0\t1\t127\t127\t1\t1.30e-61\t235\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_2455 16S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_2455\tscaffold1_cov232\t99.935\t1539\t1\t0\t1\t1539\t846845\t845307\t0.0\t2837\n",
            "TOL_2455\tscaffold1_cov232\t99.870\t1539\t2\t0\t1\t1539\t2526527\t2528065\t0.0\t2832\n",
            "TOL_2455\tscaffold32_cov590\t99.929\t1399\t1\t0\t134\t1532\t1399\t1\t0.0\t2579\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_3250 5S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_3250\tscaffold1_cov232\t100.000\t116\t0\t0\t1\t116\t841988\t841873\t5.67e-57\t215\n",
            "TOL_3250\tscaffold1_cov232\t100.000\t116\t0\t0\t1\t116\t2531384\t2531499\t5.67e-57\t215\n",
            "TOL_3250\tscaffold35_cov581\t100.000\t99\t0\t0\t1\t99\t99\t1\t1.60e-47\t183\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_3251 23S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 5 hits found\n",
            "TOL_3251\tscaffold1_cov232\t100.000\t2887\t0\t0\t1\t2887\t2528316\t2531202\t0.0\t5332\n",
            "TOL_3251\tscaffold1_cov232\t99.965\t2887\t1\t0\t1\t2887\t845056\t842170\t0.0\t5326\n",
            "TOL_3251\tscaffold35_cov581\t100.000\t1535\t0\t0\t1353\t2887\t1815\t281\t0.0\t2835\n",
            "TOL_3251\tscaffold41_cov598\t100.000\t1051\t0\t0\t215\t1265\t1051\t1\t0.0\t1941\n",
            "TOL_3251\tscaffold48_cov623\t100.000\t127\t0\t0\t1\t127\t127\t1\t1.30e-61\t235\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_3252 16S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_3252\tscaffold1_cov232\t99.935\t1539\t1\t0\t1\t1539\t846845\t845307\t0.0\t2837\n",
            "TOL_3252\tscaffold1_cov232\t99.870\t1539\t2\t0\t1\t1539\t2526527\t2528065\t0.0\t2832\n",
            "TOL_3252\tscaffold32_cov590\t99.929\t1399\t1\t0\t134\t1532\t1399\t1\t0.0\t2579\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_3253 5S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_3253\tscaffold1_cov232\t100.000\t116\t0\t0\t1\t116\t841988\t841873\t5.67e-57\t215\n",
            "TOL_3253\tscaffold1_cov232\t100.000\t116\t0\t0\t1\t116\t2531384\t2531499\t5.67e-57\t215\n",
            "TOL_3253\tscaffold35_cov581\t100.000\t99\t0\t0\t1\t99\t99\t1\t1.60e-47\t183\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_3254 23S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 5 hits found\n",
            "TOL_3254\tscaffold1_cov232\t100.000\t2887\t0\t0\t1\t2887\t2528316\t2531202\t0.0\t5332\n",
            "TOL_3254\tscaffold1_cov232\t99.965\t2887\t1\t0\t1\t2887\t845056\t842170\t0.0\t5326\n",
            "TOL_3254\tscaffold35_cov581\t100.000\t1535\t0\t0\t1353\t2887\t1815\t281\t0.0\t2835\n",
            "TOL_3254\tscaffold41_cov598\t100.000\t1051\t0\t0\t215\t1265\t1051\t1\t0.0\t1941\n",
            "TOL_3254\tscaffold48_cov623\t100.000\t127\t0\t0\t1\t127\t127\t1\t1.30e-61\t235\n",
            "# BLASTN 2.6.0+\n",
            "# Query: TOL_3255 16S ribosomal RNA\n",
            "# Database: User specified sequence set (Input: scaffolds.fasta)\n",
            "# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
            "# 3 hits found\n",
            "TOL_3255\tscaffold1_cov232\t99.935\t1539\t1\t0\t1\t1539\t846845\t845307\t0.0\t2837\n",
            "TOL_3255\tscaffold1_cov232\t99.870\t1539\t2\t0\t1\t1539\t2526527\t2528065\t0.0\t2832\n",
            "TOL_3255\tscaffold32_cov590\t99.929\t1399\t1\t0\t134\t1532\t1399\t1\t0.0\t2579\n",
            "# BLAST processed 12 queries\n"
          ]
        }
      ]
    }
  ]
}